docs/Diagnostics.md

```diff
@@ -0,0 +1,101 @@
+
+Unlike Clang and many other compilers, Swift compiler warnings cannot be disabled as part of the invocation or through a standard construct in the source. This was a deliberate choice in order to keep the language from fracturing into dialects, but has been controversial throughout Swift's history.
+
+Clang also has a kind of diagnostic called a "remark", which represents information about the build that does not indicate an issue. Swift does not currently have remarks, but there's no reason why it couldn't.
+
+
+### Grammar and Phrasing ###
+
+- Swift diagnostics should be a single phrase or sentence, with no period at the end. If it's important to include a second idea in the diagnostic, use a semicolon to separate the two parts.
+
+- Swift diagnostics do not follow strict English grammar rules. Omit words that make the diagnostic longer without adding information. This includes grammatical words like "the".
+
+- *Do* include information that shows that the compiler understands the code. For example, referring to "instance method 'foo(bar:)'" is usually unnecessary, but may increase implicit trust in the compiler and decrease the developer's mental load.
+
+  - *Don't* include information that might show that the compiler *doesn't* understand the code. For example, don't *assume* that this particular value is an instance method when it might be a property with a function type. However, *do* include details that might show how the compiler is *misunderstanding* the code.
+
+- When applicable, phrase diagnostics as rules rather than reporting that the compiler failed to do something. Example:
+
+  - Normal: "cannot call 'super.init' outside of an initializer"
+  - Better: "'super.init' cannot be called outside of an initializer"
+
+- Match the tone and phrasing of other diagnostics. Some common phrases:
+
+  - "\<noun> '\<name>'" (e.g. "type 'Set<Int>'")
+  - "value of type \<type>"
+  - "did you mean...?"
+  - "...to silence this warning"
+
+
+### Locations and Highlights ###
+
+- Diagnostics are always emitted at a particular line and column. Try to be specific.
+
+- Use highlights to build on the single location of the diagnostic, but use them sparingly. A diagnostic with *everything* on the line highlighted is no better than a diagnostic with nothing highlighted.
+
+
+### Fix-its ###
+
+- Fix-its can span multiple lines, but may not be displayed in all contexts if they do. (In particular, the command-line display of a diagnostic currently only shows the line the diagnostic is on.)
+
+- Fix-its must be in the same file as the diagnostic they are attached to. (However, a note can be in a different file than the primary error or warning.)
+
+- If a fix-it is placed on an error or warning, it must be the single, obvious, and very likely correct way to fix the issue.
+
+  - Ideally, the compiler or other tool will recover as if the user had applied the fix-it.
```

Comment:
Ah, no, that's a sub-bullet. The recovery is much less important for fix-its on notes.